src/java/test/org/apache/zookeeper/test/KeyAuthClientTest.java

```diff
@@ -0,0 +1,141 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.test;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.ZooDefs.Ids;
+import org.apache.zookeeper.ZooDefs.Perms;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.ZooKeeper;
+
+import org.apache.zookeeper.data.ACL;
+import org.apache.zookeeper.data.Id;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KeyAuthClientTest extends ClientBase {
+    private static final Logger LOG = LoggerFactory.getLogger(KeyAuthClientTest.class);
+    static final ArrayList<ACL> KeyACLs = new ArrayList<ACL>(
+                Collections.singletonList(new ACL(Perms.ALL, new Id("key","anyone"))));
+    static {
+        System.setProperty("zookeeper.authProvider.1","org.apache.zookeeper.server.auth.KeyAuthenticationProvider");
+    }
+
+    public void createNodePrintAcl(ZooKeeper zk, String path, String testName) {
+      try {
+        LOG.debug("KeyAuthenticationProvider Creating Test Node:"+path+".\n");
+        zk.create(path, null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+        List<ACL> acls = zk.getACL(path, null);
+        LOG.debug("Node: "+path+" Test:"+testName+" ACLs:");
+        for (ACL acl : acls) {
+          LOG.debug("  "+acl.toString());
+        }
+      } catch (Exception e) {
+          LOG.debug("  EXCEPTION THROWN");
+      }
+    }
+
+    public void testPreAuth() throws Exception {
+        ZooKeeper zk = createClient();
+        zk.addAuthInfo("key", "25".getBytes());
+        try {
+            createNodePrintAcl(zk, "/pre", "testPreAuth");
+            zk.setACL("/", Ids.CREATOR_ALL_ACL, -1);
+            zk.getChildren("/", false);
+            zk.create("/abc", null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+            zk.setData("/abc", "testData1".getBytes(), -1);
+            zk.create("/key", null, Ids.CREATOR_ALL_ACL, CreateMode.PERSISTENT);
+            zk.setData("/key", "5".getBytes(), -1);
+            Thread.sleep(1000);
+        } catch (KeeperException e) {
+            Assert.fail("test failed :" + e);
+        }
+        finally {
+            zk.close();
+        }
+    }
+
+    public void testMissingAuth() throws Exception {
+        ZooKeeper zk = createClient();
+        try {
+            zk.getData("/abc", false, null);
+            Assert.fail("Should not be able to get data");
+        } catch (KeeperException e) {
+        }
+        try {
+            zk.setData("/abc", "testData2".getBytes(), -1);
+            Assert.fail("Should not be able to set data");
+        } catch (KeeperException e) {
```

Comment:
I used KeyAuthClientTest.java unchanging from the original PR. I'll have a pass at it to clean things up.